Handle Sass cli options (style, --no-source_map, --embed-sources...)

smnandre opened the pull request:
Update: 

the following options are handled

```php
// Input and Output
'--style' => 'expanded',                // Output style.  [expanded (default), compressed]
'--[no-]charset' => null,               // Emit a @charset or BOM for CSS with non-ASCII characters.
'--[no-]error-css' => null,             // Emit a CSS file when an error occurs.
// Source Maps
'--[no-]source-map' => true,            // Whether to generate source maps. (defaults to on)
'--[no-]embed-sources' => null,         // Embed source file contents in source maps.
'--[no-]embed-source-map' => null,      // Embed source map contents in CSS.
// Warnings
'--[no-]quiet' => null,                 // Don't print warnings.
'--[no-]quiet-deps' => null,            // Don't print deprecation warnings for dependencies.
// Other
'--[no-]stop-on-error' => null,         // Don't compile more files once an error is encountered.
'--[no-]trace' => null,                 // Print full Dart stack traces for exceptions.
```


--- 

Been playing with this bundle a bit so ... .. is this as you imagined it ? 

I just took some options to start the work:

```
charset
style
source_map
embed_source_map
embed_sources
```

We can implement more after: https://sass-lang.com/documentation/cli/dart-sass/#no-source-map

I tried something else at first, so maybe the DTO is not really usefull

Tell me :)

stof commented on src/SassBuilder.php:
```diff
@@ -26,24 +26,26 @@ public function __construct(
         private readonly string $projectRootDir,
         private readonly ?string $binaryPath,
         private readonly bool $embedSourcemap,
+        private readonly ?array $sassOptions = [],
```
why nullable ?

smnandre replied:
Changed for a single nullable argument... don't know if this is perfect on a BC pov .. :|

stof commented on src/SassOptions.php:
```diff
@@ -0,0 +1,135 @@
+ *
+ * @see https://sass-lang.com/documentation/cli/dart-sass/
+ */
+final class SassOptions
```
This class looks useless to me, especially as it is purely internal to the SassBuilder

smnandre replied:
Removed

stof commented on src/SassBuilder.php:
```diff
@@ -26,24 +26,26 @@ public function __construct(
-        if ($this->embedSourcemap) {
-            $args[] = '--embed-source-map';
-        }
+        $args = $this->getScssCssTargets();
```
Moving this after the handling of `$watch` that was adding things in the array is broken.

stof commented on src/DependencyInjection/SymfonycastsSassExtension.php:
```diff
@@ -80,11 +86,39 @@ public function getConfigTreeBuilder(): TreeBuilder
+                            ->defaultValue(false)
+                        ->end()
                     ->end()
+                    ->ignoreExtraKeys()
```
-1 for that. Ignoring extra keys means that any mistake in the config will be silently ignored which leads to much worse DX.

stof commented on src/DependencyInjection/SymfonycastsSassExtension.php:
```diff
@@ -80,11 +86,39 @@ public function getConfigTreeBuilder(): TreeBuilder
+                            ->info('Generate a source map file (default: true)')
+                            ->defaultValue(true)
+                        ->end()
+                        ->scalarNode('embed_source_map')
```
should be a boolean node

stof commented on src/DependencyInjection/SymfonycastsSassExtension.php:
```diff
@@ -80,11 +86,39 @@ public function getConfigTreeBuilder(): TreeBuilder
+                            ->defaultValue(true)
+                        ->end()
+                        ->enumNode('style')
+                            ->info('The style of the generated CSS: compressed or expanded (default: expanded)')
```
no need to duplicate the default value in the `info` IMO (which risks becoming out of sync)

smnandre commented:
I reworked that and followed @stof  feedback 

Open to any comment / suggestions :)

bocharsky-bw commented on src/SassBuilder.php:
```diff
@@ -18,29 +18,58 @@ class SassBuilder
+        'source_map' => true,
+        'embed_source_map' => false,
+        'embed_sources' => false,
+    ];
```
Hm, do we need this? It seems like we will have to keep this array in sync with default values in the `SymfonycastsSassExtension` and it's easy to mismatch, i.e. change the defaults in one spot only that will cause issues

smnandre replied:
It's a bit more than that, and my code lacks needed comments it seems :)

Those are the default "sass cli" options.

So they can be different from what we do want as default options. 

Or maybe i'm over-thinking all this.

bocharsky-bw replied:
Hm, but sass cli defaults are already defaults upstream, so if we will not pass options to the sass cli they will be defaulted. Or am I missing something?

smnandre replied:
The defaults from the bundle depends on the debug (and they probably could/should depend on the env too). I just wanted a place to store the default CLI values, but if you want, i remove it no problem :) 

Do i still keep some "keys" there ... or do we allow any key/value ?

bocharsky-bw replied:
Ah, I probably see the reason now, thanks for explaining!

bocharsky-bw commented on src/SassBuilder.php:
```diff
@@ -97,6 +126,17 @@ public static function guessCssNameFromSassFile(string $sassFile, string $output
+            throw new \InvalidArgumentException(sprintf('Invalid Sass options: "%s". Valid options are: "%s".', implode('", "', $invalidOptions), implode('", "', $validOptions)));
+        }
+
+        return array_diff_assoc($options, self::DEFAULT_OPTIONS);
```
So here we return values only different from the defaults in `self::DEFAULT_OPTIONS`?

smnandre commented on src/SassBuilder.php:
```diff
@@ -18,29 +18,58 @@ class SassBuilder
     ) {
+        if (\is_bool($sassOptions)) {
+            // Until 0.3, the $sassOptions argument was a boolean named $embedSourceMap
+            trigger_deprecation('symfonycasts/sass-bundle', '0.3', 'Passing a boolean to embed the source map is deprecated. Use [\'embed_source_map\' => true] instead.');
```
Yes thank you! 

(I'll fix it with the 0.3 so i won't use your commit suggestion as it)

smnandre commented:
How would you like to advance on this topic ?

bocharsky-bw commented on src/DependencyInjection/SymfonycastsSassExtension.php:
```diff
@@ -27,11 +27,15 @@ public function load(array $configs, ContainerBuilder $container): void
 
+        if (isset($config['embed_sourcemap'])) {
+            $config['sass_options']['embed_source_map'] = $config['embed_sourcemap'];
+        }
```
Does it worth mentioning that it's here for BC?

smnandre replied:
I’ll add some comments.. and update the doc :)

bocharsky-bw reviewed (approved):
Sorry for the delay!

Yeah, it seems like this one is ready for me, I don't have much to add, though hoped to hear some other opinions on it

smnandre commented:
I'm reworking this PR (just a bit) to make the options more obvious for everyone, maybe tonight, tomorrow elsewhen.

smnandre commented:
So, i reworked how the options are handled in what i personnaly find is more simple to handle.

1) I removed the default values from the dependency injection / configuration
2) The management of the [no] is internal in the Builder and - i find - more easy to manage / understand.

The 2 option format (php in snake case, CLI in --[no-]dashed) are mapped in the builder

I'd just need some help to add some of the missing tests (deps, errors, ...) .. or if you can wait a bit, i will have time in a couple of days.

bocharsky-bw commented:
Thanks! No rush at all, we can definitely wait 👍

smnandre commented:
That feels OK for me.. open to any feedback!

And if anyone wants to check on its side please do, i don't have any big sass project on my radar right now to fully test "live"

bocharsky-bw reviewed (approved):
Hey Simon!

Thank you for rewriting it! It seems more solid now, so I think it's ready 👍

smnandre commented on src/SassBuilder.php:
```diff
@@ -97,8 +128,76 @@ public static function guessCssNameFromSassFile(string $sassFile, string $output
+                continue;
+            }
+            // --style=compressed
+            if (\is_string($value)) {
```
And call it like this from the build command: 

```
  $args = [
      ...$this->getScssCssTargets(),
      ...$this->getBuildOptions(['--watch' => $watch]),
  ];
```

That will also allow us to change the signature (later) from `build($watch)` to `build(arra $options) ` and so we could expose **some** options in the BuildCommand 

WDYT ?

smnandre replied:
I pushed this idea, i can revert anytime obviously :)

bocharsky-bw replied:
Wait, not sure it will work... as `--watch` value IS a string, i.e. we will fall into that `if (\is_string($value)) {` and do something like `--watch=--watch` what we don't need, right? it seems the previous solution was better

smnandre replied:
https://github.com/SymfonyCasts/sass-bundle/pull/39/files

It's now 

```php
...$this->getBuildOptions(['--watch' => $watch]),
```

So that should work as before, but in a more "generic" way

smnandre commented:
I have no idea how you'd imagine the documentation so feel free to tell me if what i added is not in the spirit of this Bundle :)

bocharsky-bw commented on doc/index.rst:
```diff
@@ -140,8 +140,57 @@ To see the full config from this bundle, run:
+        root_sass:  'assets/styles/app.scss'
+
+
+Sass Cli Options
```
```suggestion
Sass CLI Options
```

bocharsky-bw commented on doc/index.rst:
```diff
@@ -140,8 +140,57 @@ To see the full config from this bundle, run:
+Sass Cli Options
+~~~~~~~~~~~~~~~~
+
+You can configure most of the `Dart Sass cli options <https://sass-lang.com/documentation/cli/dart-sass>`_:
```
```suggestion
You can configure most of the `Dart Sass CLI options <https://sass-lang.com/documentation/cli/dart-sass>`_:
```

bocharsky-bw commented:
About the docs - list all the possible options is just fine for me

smnandre commented:
Hmm something has no worked / i messed my push ...

`--watch` should be the option and "true" the value

smnandre commented:
It is 

```php
...$this->getBuildOptions(['--watch' => $watch]),
```

I'm not sure if something happened with the git log (possible as i have this bad habit to destroy git histories on a daily basis 🤒 )

or maybe github did not show you the "flatten" changes ?

bocharsky-bw reviewed (approved):
Ah, it might be so that I just missed it... or reviewed not full change set. Makes sense to me now! 👍 

OK, I think we're completely ready here, thank you for your work on it!

smnandre commented:
It's been a pleasure...  😄  

But don't thank me too soon, we'll see how many complaints/bugs you'll receive haha
